pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,470 @@
+	if err != nil {
+		return err
+	}
+
+	// Create or update the latency store
+	mutator := func(entry *NodeIPLatencyEntry) {
+		entry.LastSendTime = timeStart
+	}
+	m.latencyStore.SetNodeIPLatencyEntry(addr.String(), mutator)
+
+	return nil
+}
+
+// recvPing receives an ICMP message from the target IP address.
+func (m *NodeLatencyMonitor) recvPing(socket net.PacketConn, isIPv4 bool) {
+	// We only expect small packets, if we receive a larger packet, we will drop the extra data.
+	readBuffer := make([]byte, 128)
+	for {
+		n, peer, err := socket.ReadFrom(readBuffer)
+		if err != nil {
+			// When the socket is closed in the Run method, this error will be logged, which is not ideal.
+			// In the future, we may try setting a ReadDeadline on the socket before each ReadFrom and using
+			// a channel to signal that the loop should terminate.
+			klog.ErrorS(err, "Failed to read ICMP message")
+			return
+		}
+
+		destIP := peer.String()
+
+		// Parse the ICMP message
+		var msg *icmp.Message
+		if isIPv4 {
+			msg, err = icmp.ParseMessage(ProtocolICMP, readBuffer[:n])
+			if err != nil {
+				klog.ErrorS(err, "Failed to parse ICMP message")
+				continue
+			}
+			if msg.Type != ipv4.ICMPTypeEcho && msg.Type != ipv4.ICMPTypeEchoReply {
+				klog.V(5).InfoS("Ignoring non-ping ICMP message", "msg", msg)
+				continue
+			}
+			// Ignore ICMP echo messages received from other Nodes (they will be answered by the system)
+			if msg.Type == ipv4.ICMPTypeEcho {
+				klog.V(7).InfoS("Ignoring ICMP echo request message", "msg", msg)
+				continue
+			}
+		} else {
+			msg, err = icmp.ParseMessage(ProtocolICMPv6, readBuffer)
+			if err != nil {
+				klog.ErrorS(err, "Failed to parse ICMP message")
+				continue
+			}
+			if msg.Type != ipv6.ICMPTypeEchoRequest && msg.Type != ipv6.ICMPTypeEchoReply {
+				klog.V(5).InfoS("Ignoring non-ping ICMP message", "msg", msg)
+				continue
+			}
+			// Ignore ICMP echo messages received from other Nodes (they will be answered by the system)
+			if msg.Type == ipv6.ICMPTypeEchoRequest {
+				klog.V(7).InfoS("Ignoring ICMP echo request message", "msg", msg)
+				continue
+			}
+		}
+
+		echo, ok := msg.Body.(*icmp.Echo)
+		if !ok {
+			klog.ErrorS(nil, "Failed to assert type as *icmp.Echo")
+			continue
+		}
+		if echo.ID != icmpEchoID {
+			klog.V(4).InfoS("Ignoring ICMP message with wrong echo ID", "msg", msg)
+		}
+
+		klog.V(4).InfoS("Recv ICMP message", "IP", destIP, "msg", msg)
+
+		// Parse the time from the ICMP data
+		sentTime, err := time.Parse(time.RFC3339Nano, string(echo.Data))
+		if err != nil {
+			klog.ErrorS(err, "Failed to parse time from ICMP data")
+			continue
+		}
+
+		// Calculate the round-trip time
+		end := time.Now()
+		rtt := end.Sub(sentTime)
+		klog.V(4).InfoS("Updating latency entry for Node IP", "IP", destIP, "lastSendTime", sentTime, "lastRecvTime", end, "RTT", rtt)
+
+		// Update the latency store
+		mutator := func(entry *NodeIPLatencyEntry) {
+			entry.LastSendTime = sentTime
```

Comment:
It seems LastSendTime could be changed from recorded time to the timestemp stored in the payload, though they should be the same one. Maybe not a big deal, but it seems setting it when sending the request looks unnecessary, or we should use it to validate if the response matches? @antoninbas @Dyanngg @heanlan

Reply:
I may be misunderstanding the question, but we want to keep increasing  `LastSendTime` if connectivity is broken and no response has been received in a while.

Reply:
OK, I was wondering the point of setting LastSendTime when sending the request while overriding it when receiving the response. Now I get that increasing it is definitely helpful to tell how long it has been receiving no response. However, should receiving a response override it? It seems to cause the meaning of the timestamp to change between "the timestamp of sending the last request" and "the timestamp of sending the request of the last response", and it could even decrease in some extreme cases, like the following case:
At t1, it sends REQ1, LastSendTime is set to t1;
At t2, it sends REQ2, LastSendTime is set to t2;
At t3, it receives RESP1, LastSendTime is set to t1, LastRecvTime is set to t3.

Reply:
We should not override it when processing the response. This is indeed not what was intended.

Reply:
That's right, I will ignore to update send time in `recv`.